### PR TITLE
peek

### PR DESCRIPTION
**Pull request by elacuesta**
- [x] `queuelib.queue`
- [x] `queuelib.pqueue`
- [x] `queuelib.rrqueue`

### CODE REVIEW

**Comment by elacuesta** [comment]
The only line not covered by the tests is one that is also not covered in the current `pop` methods for disk queues, and I'm not sure how to cover it. It might involve corrupting the underlying file?

File: queuelib/queue.py [hunk 1]
```diff
@@ -88,6 +94,20 @@ def pop(self):
...
 
+    def peek(self):
+        tnum, tcnt, toffset = self.info['tail']
+        if [tnum, tcnt] >= self.info['head']:
+            return
+        tfd = self.tailf.fileno()
+        tfd_initial_pos = os.lseek(tfd, 0, os.SEEK_CUR)
+        szhdr = os.read(tfd, self.szhdr_size)
+        if not szhdr:
+            return
```
Comment (Gallaecio) [inline queuelib/queue.py:105]: I wonder why we check the output of `os.read` here but not a few lines below.

I’m not 100% sure, but maybe it would make sense to expect the file _not_ to be corrupt, and let code fail naturally if it is corrupted.

Comment (elacuesta) [reply queuelib/queue.py:105]: I "mocked" (in quotes because I didn't actually use `unittest.mock`) one of the underlying files and that allowed me to reach the affected lines: 4db96d0b0c5aaf2e1ace8144fd248e81a1afe9ae

Comment (Gallaecio) [reply queuelib/queue.py:105]: My point was that the likelihood of this breaking and the `os.read` below breaking is not that different, but here we check, and below we don’t. It feels inconsistent.

Should we have this below as well for consistency?

```
data = os.read(tfd, size)
if not data:
    return
```

The result would be slightly different, causing the scenario of `tfd` being empty to return `None` instead of `b''`.

However, my point was that `if not szhdr:` seems to only matter if something unexpected happened to the file, out of the control of queuelib. Is that something we should worry about? Can’t we just let things break loudly if something out of the control of queuelib corrupts the queue file?

In other words, instead of mocking to cover those lines with tests, wouldn’t it be OK to remove the `if not szhdr:` part altogether?

Comment (elacuesta) [reply queuelib/queue.py:105]: Ok, I see what you mean. I think it's a valid concern, but to be honest, I don't know for sure why the check was included in the first place. I tend to agree that we should probably fail loudly, but I don't feel confident enough to fully support that without a deeper understanding of this part of the code. That said, I don't think it should be a blocker for this PR. I think it's consistent to do it in `peek` if we do it in `pop`, and we can open a new issue to keep track of it separately.

**Review by Gallaecio — APPROVED** [review]

**Comment by easyteacher** [comment]
This PR is likely to cause scrapy fails to pass tests. Can you look into it?

> [  372s] =================================== FAILURES ===================================
> [  372s] ___________________ MarshalFifoDiskQueueTest.test_peek_fifo ____________________
> [  372s] 
> [  372s] self = <tests.test_squeues.MarshalFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ________________ MarshalFifoDiskQueueTest.test_peek_one_element ________________
> [  372s] 
> [  372s] self = <tests.test_squeues.MarshalFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] ______________ ChunkSize1MarshalFifoDiskQueueTest.test_peek_fifo _______________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize1MarshalFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ___________ ChunkSize1MarshalFifoDiskQueueTest.test_peek_one_element ___________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize1MarshalFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] ______________ ChunkSize2MarshalFifoDiskQueueTest.test_peek_fifo _______________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize2MarshalFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ___________ ChunkSize2MarshalFifoDiskQueueTest.test_peek_one_element ___________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize2MarshalFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] ______________ ChunkSize3MarshalFifoDiskQueueTest.test_peek_fifo _______________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize3MarshalFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ___________ ChunkSize3MarshalFifoDiskQueueTest.test_peek_one_element ___________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize3MarshalFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] ______________ ChunkSize4MarshalFifoDiskQueueTest.test_peek_fifo _______________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize4MarshalFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ___________ ChunkSize4MarshalFifoDiskQueueTest.test_peek_one_element ___________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize4MarshalFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] ____________________ PickleFifoDiskQueueTest.test_peek_fifo ____________________
> [  372s] 
> [  372s] self = <tests.test_squeues.PickleFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ________________ PickleFifoDiskQueueTest.test_peek_one_element _________________
> [  372s] 
> [  372s] self = <tests.test_squeues.PickleFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] _______________ ChunkSize1PickleFifoDiskQueueTest.test_peek_fifo _______________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize1PickleFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ___________ ChunkSize1PickleFifoDiskQueueTest.test_peek_one_element ____________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize1PickleFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] _______________ ChunkSize2PickleFifoDiskQueueTest.test_peek_fifo _______________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize2PickleFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ___________ ChunkSize2PickleFifoDiskQueueTest.test_peek_one_element ____________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize2PickleFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] _______________ ChunkSize3PickleFifoDiskQueueTest.test_peek_fifo _______________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize3PickleFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ___________ ChunkSize3PickleFifoDiskQueueTest.test_peek_one_element ____________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize3PickleFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] _______________ ChunkSize4PickleFifoDiskQueueTest.test_peek_fifo _______________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize4PickleFifoDiskQueueTest testMethod=test_peek_fifo>
> [  372s] 
> [  372s]     def test_peek_fifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:159: AssertionError
> [  372s] ___________ ChunkSize4PickleFifoDiskQueueTest.test_peek_one_element ____________
> [  372s] 
> [  372s] self = <tests.test_squeues.ChunkSize4PickleFifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] ___________________ MarshalLifoDiskQueueTest.test_peek_lifo ____________________
> [  372s] 
> [  372s] self = <tests.test_squeues.MarshalLifoDiskQueueTest testMethod=test_peek_lifo>
> [  372s] 
> [  372s]     def test_peek_lifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"c")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00c' != b'c'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:203: AssertionError
> [  372s] ________________ MarshalLifoDiskQueueTest.test_peek_one_element ________________
> [  372s] 
> [  372s] self = <tests.test_squeues.MarshalLifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\xf3\x01\x00\x00\x00a' != b'a'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:122: AssertionError
> [  372s] ____________________ PickleLifoDiskQueueTest.test_peek_lifo ____________________
> [  372s] 
> [  372s] self = <tests.test_squeues.PickleLifoDiskQueueTest testMethod=test_peek_lifo>
> [  372s] 
> [  372s]     def test_peek_lifo(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s]         q.push(b"b")
> [  372s]         q.push(b"c")
> [  372s] >       self.assertEqual(q.peek(), b"c")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01c\x94.' != b'c'
> [  372s] 
> [  372s] /usr/lib/python3.9/site-packages/queuelib/tests/test_queue.py:203: AssertionError
> [  372s] ________________ PickleLifoDiskQueueTest.test_peek_one_element _________________
> [  372s] 
> [  372s] self = <tests.test_squeues.PickleLifoDiskQueueTest testMethod=test_peek_one_element>
> [  372s] 
> [  372s]     def test_peek_one_element(self):
> [  372s]         q = self.queue()
> [  372s]         self.assertIsNone(q.peek())
> [  372s]         q.push(b"a")
> [  372s] >       self.assertEqual(q.peek(), b"a")
> [  372s] E       AssertionError: b'\x80\x04\x95\x05\x00\x00\x00\x00\x00\x00\x00C\x01a\x94.' != b'a'

**Comment by easyteacher** [comment]
> Scrapy and its tests were updated to support `peek` at [scrapy/scrapy#5112](https://github.com/scrapy/scrapy/pull/5112). What version of the tests are you running?

v2.5.0. Now the patch is applied and thank you for mentioning.

https://build.opensuse.org/package/view_file/devel:languages:python/python-Scrapy/add-peak-method-to-queues.patch?expand=1
